komposable-architecture-test/src/test/java/com/toggl/komposable/test/TestStoreTests.kt

```diff
@@ -0,0 +1,403 @@
+package com.toggl.komposable.test
+
+import com.toggl.komposable.architecture.Effect
+import com.toggl.komposable.architecture.ReduceResult
+import com.toggl.komposable.architecture.Reducer
+import com.toggl.komposable.extensions.cancellable
+import com.toggl.komposable.extensions.merge
+import com.toggl.komposable.extensions.withCancelEffect
+import com.toggl.komposable.extensions.withEffect
+import com.toggl.komposable.extensions.withSuspendEffect
+import com.toggl.komposable.extensions.withoutEffect
+import com.toggl.komposable.scope.DispatcherProvider
+import com.toggl.komposable.test.store.TestStore.Exhaustivity
+import com.toggl.komposable.test.store.createTestStore
+import io.kotest.assertions.throwables.shouldNotThrow
+import io.kotest.assertions.throwables.shouldThrow
+import io.kotest.assertions.withClue
+import io.kotest.matchers.shouldBe
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.test.StandardTestDispatcher
+import kotlinx.coroutines.test.TestScope
+import kotlinx.coroutines.test.advanceTimeBy
+import kotlinx.coroutines.test.runTest
+import kotlinx.coroutines.test.setMain
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Nested
+import org.junit.jupiter.api.Test
+import kotlin.time.Duration
+
+class TestStoreTests {
+    open class BaseTestStoreTest {
+        private val testDispatcher = StandardTestDispatcher()
+        val dispatcherProvider = DispatcherProvider(testDispatcher, testDispatcher, testDispatcher)
+        val testCoroutineScope = TestScope(testDispatcher)
+
+        @BeforeEach
+        fun beforeTest() {
+            Dispatchers.setMain(testDispatcher)
+        }
+    }
+    private enum class MergeTestsAction {
+        A, B1, B2, B3, C1, C2, C3, D
+    }
+
+    @Nested
+    inner class MergeTests : BaseTestStoreTest() {
+        @Test
+        fun `merging working`() = runTest {
```

Comment:
The name of this test is quite vague 🤔 It's not even clear what you are merging.

Reply:
this is a place where we differ from TCA, that's our different behaviour (TCA's `merge` apparently works like their `concatenate`).